File: Hard/permutation_sequence/permutation_sequence.py
from typing import List

def getPermutation(n: int, k: int) -> str:
    # Create a list of numbers to get permutations from
    numbers = list(range(1, n + 1))
    # Calculate the factorial of (n-1) to determine the block size
    factorial = 1
    for i in range(1, n):
        factorial *= i

    # Adjust k to be zero-indexed
    k -= 1

    result = []
    for i in range(n):
        index = k // factorial
        result.append(str(numbers[index]))
        # Remove the used number from the list
        numbers.pop(index)

        if i < n - 1:
            k %= factorial
            factorial //= (n - 1 - i)

    return "".join(result)

if __name__ == "__main__":
    print("Testing getPermutation function:")
    print("For n = 3 and k = 3, the output is:", getPermutation(3, 3)) #Output: "213"
    print("For n = 4 and k = 9, the output is:", getPermutation(4, 9)) #Output: "2314"
    print("For n = 3 and k = 1, the output is:", getPermutation(3, 1)) #Output: "123"

File: Hard/permutation_sequence/permutation_sequence.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
    public:
        string getPermutation(int n, int k) {
            vector<int> factorial(n, 1);
            for(int i = 1; i < n; i++) {
                factorial[i] = factorial[i-1] * i;
            }
            vector<int> numbers;
            for(int i = 1; i <= n; i++) {
                numbers.push_back(i);
            }
            k--; // Convert to zero-based index
            string result;
            for(int i = n - 1; i >= 0; i--) {
                int index = k / factorial[i];
                result += to_string(numbers[index]);
                numbers.erase(numbers.begin() + index); // Remove used number
                k %= factorial[i];
            }
            return result;
        }
};

// Example usage:
int main() {
    Solution solution1;
    int n1 = 3, k1 = 3;
    string result1 = solution1.getPermutation(n1, k1);
    cout << "Testing Permutation Sequence: " << endl;
    cout << "Input: n = 3, k = 3" << endl;
    cout << "Output: " << result1 << endl;

    cout << "" << endl;
    Solution solution2;
    int n2 = 4, k2 = 9;
    string result2 = solution2.getPermutation(n2, k2);
    cout << "Input: n = 4, k = 9" << endl;
    cout << "Output: " << result2 << endl;

    cout << "" << endl;
    Solution solution3;
    int n3 = 3, k3 = 1;
    string result3 = solution3.getPermutation(n3, k3);
    cout << "Input: n = 3, k = 1" << endl;
    cout << "Output: " << result3 << endl;

    return 0;
}